internal/vulnerability/vulnerability_updater.go

```diff
@@ -0,0 +1,233 @@
+	return int64(lastOccurrenceDate.Sub(firstOccurrenceDate).Hours() / 24)
+}
+
+func firstOccurrenceToLastParams(metric *dependencytrack.ProjectMetric, projectID uuid.UUID) []gensql.VulnerabilityMetricsUpsertParams {
+	var vulnerabilityMetricsUpsertParams []gensql.VulnerabilityMetricsUpsertParams
+	for _, vulnSum := range metric.VulnerabilityMetrics {
+		firstOccurrenceDateTime := toDateTime(vulnSum.FirstOccurrence)
+		lastOccurrenceDateTime := toDateTime(vulnSum.LastOccurrence)
+		daysToUpsert := differenceInDaysFrom(firstOccurrenceDateTime, lastOccurrenceDateTime)
+		for i := daysToUpsert + 0; i > 0; i-- {
+			vulnerabilityMetricsUpsertParams = append(vulnerabilityMetricsUpsertParams, vulnParams(vulnSum, firstOccurrenceDateTime, projectID))
+			firstOccurrenceDateTime = firstOccurrenceDateTime.AddDate(0, 0, 1)
+		}
+	}
+	return vulnerabilityMetricsUpsertParams
+}
+
+func dbLastToCurrentTimeParams(metric *dependencytrack.ProjectMetric, dbLastDateTime time.Time, currentTime time.Time, projectID uuid.UUID) []gensql.VulnerabilityMetricsUpsertParams {
+	var vulnerabilityMetricsUpsertParams []gensql.VulnerabilityMetricsUpsertParams
+	daysToUpsert := differenceInDaysFrom(dbLastDateTime, currentTime)
+	currentVulnSum := metric.VulnerabilityMetrics[len(metric.VulnerabilityMetrics)-1]
+	if currentVulnSum != nil {
+		for i := daysToUpsert + 0; i > 0; i-- {
+			dbLastDateTime = dbLastDateTime.AddDate(0, 0, 1)
+			vulnerabilityMetricsUpsertParams = append(vulnerabilityMetricsUpsertParams, vulnParams(currentVulnSum, dbLastDateTime, projectID))
+		}
+	}
+	return vulnerabilityMetricsUpsertParams
+}
+
+func (u *Updater) vulnerabilityMetricsParams(metric *dependencytrack.ProjectMetric, dbLastDateTime time.Time, currentTime time.Time) []gensql.VulnerabilityMetricsUpsertParams {
+	var vulnerabilityMetricsUpsertParams []gensql.VulnerabilityMetricsUpsertParams
+	projectID := metric.ProjectID
+
+	// If the last date is not zero, we fetch all metrics from the last occurrence to the current date
+	if !dbLastDateTime.IsZero() {
+		vulnerabilityMetricsUpsertParams = dbLastToCurrentTimeParams(metric, dbLastDateTime, currentTime, projectID)
+		return vulnerabilityMetricsUpsertParams
+	}
+
+	// If the last date is zero, we fetch all metrics from the first occurrence to the last occurrence
+	vulnerabilityMetricsUpsertParams = firstOccurrenceToLastParams(metric, projectID)
+	return vulnerabilityMetricsUpsertParams
+}
+
+func setHistoricalDate(date time.Time) string {
+	if date.IsZero() {
+		return DependencyTrackHistoricalDate
+	}
+	return date.Format("2006-01-02")
+}
+
+func (u *Updater) UpsertBatchParams(ctx context.Context, teamSlug slug.Slug, appInstances []*dependencytrack.AppInstance, currentTime time.Time) ([]gensql.VulnerabilityMetricsUpsertParams, error) {
+	var wg sync.WaitGroup
+	now := time.Now()
+	vulnerabilityMetricsUpsertParams := make([]gensql.VulnerabilityMetricsUpsertParams, 0)
+	for _, app := range appInstances {
+		wg.Add(1)
```

Comment:
Bruk 3 min på å se om denne kan redusere maks last: https://github.com/sourcegraph/conc
Dersom vi har teams med 100+ apper, så vil database connections spesielt henge for brukerne som besøker siden/bruker api-et.

Reply:
Var ikke så kritisk som først antatt, ser ut til at vi har 100 max :innocent: